app/src/main/java/org/thoughtcrime/securesms/contacts/sync/ContactDiscoveryV2.java

```diff
@@ -0,0 +1,113 @@
+
+  @WorkerThread
+  static DirectoryResult getDirectoryResult(@NonNull Context context,
+                                            @NonNull Set<String> databaseNumbers,
+                                            @NonNull Set<String> systemNumbers)
+      throws IOException
+  {
+    Set<String>                        allNumbers        = SetUtil.union(databaseNumbers, systemNumbers);
+    FuzzyPhoneNumberHelper.InputResult inputResult       = FuzzyPhoneNumberHelper.generateInput(allNumbers, databaseNumbers);
+    Set<String>                        sanitizedNumbers  = sanitizeNumbers(inputResult.getNumbers());
+    Set<String>                        ignoredNumbers    = new HashSet<>();
+
+    if (sanitizedNumbers.size() > MAX_NUMBERS) {
+      Set<String> randomlySelected = randomlySelect(sanitizedNumbers, MAX_NUMBERS);
+
+      ignoredNumbers   = SetUtil.difference(sanitizedNumbers, randomlySelected);
+      sanitizedNumbers = randomlySelected;
+    }
+
+    SignalServiceAccountManager accountManager = ApplicationDependencies.getSignalServiceAccountManager();
+    KeyStore                    iasKeyStore    = getIasKeyStore(context);
+
+    try {
+      Map<String, ACI>                    results      = accountManager.getRegisteredUsers(iasKeyStore, sanitizedNumbers, BuildConfig.CDS_MRENCLAVE);
+      FuzzyPhoneNumberHelper.OutputResult outputResult = FuzzyPhoneNumberHelper.generateOutput(results, inputResult);
+
+      return new DirectoryResult(outputResult.getNumbers(), outputResult.getRewrites(), ignoredNumbers);
+    } catch (SignatureException | UnauthenticatedQuoteException | UnauthenticatedResponseException | Quote.InvalidQuoteFormatException |InvalidKeyException e) {
+      Log.w(TAG, "Attestation error.", e);
+      throw new IOException(e);
+    }
+  }
+
+  static @NonNull DirectoryResult getDirectoryResult(@NonNull Context context, @NonNull String number) throws  IOException {
+    return getDirectoryResult(context, Collections.singleton(number), Collections.singleton(number));
+  }
+
+  private static Set<String> sanitizeNumbers(@NonNull Set<String> numbers) {
+    return Stream.of(numbers).filter(number -> {
+      try {
+        return number.startsWith("+") && number.length() > 1 && number.charAt(1) != '0' && Long.parseLong(number.substring(1)) > 0;
+      } catch (NumberFormatException e) {
+        return false;
+      }
+    }).collect(Collectors.toSet());
+  }
+
+  private static @NonNull Set<String> randomlySelect(@NonNull Set<String> numbers, int max) {
+    List<String> list = new ArrayList<>(numbers);
+    Collections.shuffle(list);
+
+    return new HashSet<>(list.subList(0, max));
+  }
+
+  private static KeyStore getIasKeyStore(@NonNull Context context) {
+    try {
+      TrustStore contactTrustStore = new IasTrustStore(context);
+
+      KeyStore keyStore = KeyStore.getInstance("BKS");
+      keyStore.load(contactTrustStore.getKeyStoreInputStream(), contactTrustStore.getKeyStorePassword().toCharArray());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource returned by the following method call: *getKeyStoreInputStream*. Currently, there are execution paths that do not contain closure statements, for example, when KeyStore.load throws an exception. Ensure the object returned by *getKeyStoreInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).